File: src/StudentRoadMap.Domain/Identity/AuditActions.cs
namespace StudentRoadMap.Domain.Identity;

/// <summary>
/// `AuditLog.Action` uchun qat'iy qiymatlar ro'yxati — `docs/08-auth-va-xavfsizlik.md`
/// 8-bo'limidagi auth bilan bog'liq harakatlar. `Auth.TotpDisabled` — `EnableTotp` bilan
/// simmetrik (`docs/13-auth-va-jwt.md` DisableTotp use-case'i talab qiladi), PM tomonidan
/// `docs/08`ga rasman kiritilgan.
/// </summary>
public static class AuditActions
{
    public const string AuthLoginSucceeded = "Auth.LoginSucceeded";
    public const string AuthLoginFailed = "Auth.LoginFailed";
    public const string AuthPasswordChanged = "Auth.PasswordChanged";
    /// <summary>
    /// `POST /api/auth/totp/enable` — o'rnatish BOSHLANDI (sir kutish holatida saqlandi).
    /// P46da qo'shildi: `Auth.TotpEnabled` endi faqat `confirm` muvaffaqiyatli tugaganda,
    /// ya'ni 2FA HAQIQATDAN yoqilganda yoziladi. Sirning o'zi hech qachon yozilmaydi.
    /// </summary>
    public const string AuthTotpEnrollmentStarted = "Auth.TotpEnrollmentStarted";

    public const string AuthTotpEnabled = "Auth.TotpEnabled";
    public const string AuthTotpDisabled = "Auth.TotpDisabled";
    public const string SecurityRefreshReuse = "Security.RefreshReuse";

    // --- P14 (`prompts/14-admin-maktab-va-oquvchi-api.md`) — `docs/08` 8-bo'limida ro'yxat
    // qilingan maktab/o'quvchi harakatlari. ---

    public const string SchoolCreated = "School.Created";
    public const string SchoolUpdated = "School.Updated";
    public const string SchoolDeleted = "School.Deleted";
    public const string SchoolLinkRegenerated = "School.LinkRegenerated";
    public const string SchoolToggledActive = "School.ToggledActive";
    public const string StudentDeleted = "Student.Deleted";

    // --- P15 (`prompts/15-sessiya-va-dashboard-api.md`) — `docs/05` §8 / `docs/08` §8da
    // ro'yxat qilingan sessiya harakatlari. `Assessment.AnalysisRerun` bu yerda YO'Q edi — u
    // `rerun-analysis` (AI, P18) uchun, P15 qamroviga kirmagan edi; P18da shu yerga qo'shildi. ---

    public const string AssessmentDeleted = "Assessment.Deleted";
    public const string AssessmentScoresRecalculated = "Assessment.ScoresRecalculated";

    // --- P18 (`prompts/18-ai-navbat-va-orkestratsiya.md`) — AI qayta tahlil va sozlamalar
    // harakatlari (`docs/07` §3.3/§3.5, `CLAUDE.md` MAXSUS DIQQAT #6: "Audit: AiConfig.Updated,
    // AiConfig.KeyChanged (kalit qiymati emas!)"). ---

    /// <summary>`POST /api/admin/assessments/{id}/rerun-analysis`.</summary>
    public const string AssessmentAnalysisRerun = "Assessment.AnalysisRerun";

    /// <summary>`PUT /api/admin/ai/providers/{provider}` — model/limit/faollik/tartib o'zgarganda (kalit o'zgarmasa ham).</summary>
    public const string AiConfigUpdated = "AiConfig.Updated";

    /// <summary>
    /// `PUT /api/admin/ai/providers/{provider}` ICHIDA, faqat API kalitining o'zi almashtirilganda
    /// QO'SHIMCHA ravishda (`AiConfigUpdated` bilan BIRGA) yoziladi — kalit QIYMATI hech qachon
    /// audit yozuvida bo'lmaydi, faqat "o'zgardi" fakti.
    /// </summary>
    public const string AiConfigKeyChanged = "AiConfig.KeyChanged";

    // --- P34 (`prompts/34-dastur-modeli-va-biriktirish.md`) — `AssessmentProgram` admin
    // harakatlari (E16-band ro'yxati: "Program.Created/Updated/Published/Archived/Assigned/
    // Unassigned"). ---

    public const string ProgramCreated = "Program.Created";
    public const string ProgramUpdated = "Program.Updated";
    public const string ProgramPublished = "Program.Published";
    public const string ProgramArchived = "Program.Archived";
    public const string ProgramRestored = "Program.Restored";
    public const string ProgramToggledActive = "Program.ToggledActive";
    public const string ProgramTestAdded = "Program.TestAdded";
    public const string ProgramTestRemoved = "Program.TestRemoved";
    public const string ProgramTestsReordered = "Program.TestsReordered";
    public const string ProgramSchoolAssigned = "Program.Assigned";
    public const string ProgramSchoolUnassigned = "Program.Unassigned";

    // --- P37 (`prompts/37-katalog-crud-backend.md`) — test katalogi CRUD admin harakatlari
    // (vazifa 11-band ro'yxati). ---

    public const string CatalogTestCreated = "Catalog.TestCreated";
    public const string CatalogTestUpdated = "Catalog.TestUpdated";
    public const string CatalogTestPublished = "Catalog.TestPublished";
    public const string CatalogTestArchived = "Catalog.TestArchived";
    public const string CatalogTestDeleted = "Catalog.TestDeleted";
    public const string CatalogTestDuplicated = "Catalog.TestDuplicated";
    public const string CatalogTestToggledActive = "Catalog.TestToggledActive";
    public const string CatalogScaleChanged = "Catalog.ScaleChanged";
    public const string CatalogQuestionAdded = "Catalog.QuestionAdded";
    public const string CatalogQuestionRemoved = "Catalog.QuestionRemoved";
    public const string CatalogQuestionUpdated = "Catalog.QuestionUpdated";
    public const string CatalogVersionBumped = "Catalog.VersionBumped";

    // --- P27 (`prompts/27-eksport-excel-va-pdf.md`) — `docs/08` §8 ro'yxatida bor. ---

    /// <summary>O'quvchilar ro'yxati `.xlsx` eksporti (`docs/08` §8) — kim, qachon, qaysi filtr, nechta qator.</summary>
    public const string ExportStudentsDownloaded = "Export.StudentsDownloaded";
}

File: frontend/src/features/audit/model/auditActions.ts
/**
 * Ma'lum harakat/obyekt turlari — filtr select'lari uchun (`docs/11` A-9: "filtr: harakat
 * turi, obyekt, sana"). `GET /api/admin/audit-logs?action=&entityType=` ANIQ moslikda ishlaydi
 * (`ListAuditLogsQueryHandler` izohi), shu sabab erkin matn emas — qat'iy ro'yxat.
 *
 * Ro'yxat ikki manbadan birlashtirilgan:
 * 1. `src/StudentRoadMap.Domain/Identity/AuditActions.cs` — HOZIR real yoziladigan qiymatlar
 *    (Auth/School/Student/Assessment/Program).
 * 2. `docs/08-auth-va-xavfsizlik.md` §8 — to'liq REJALASHTIRILGAN ro'yxat, shu jumladan
 *    `AiConfig.*`/`Catalog.QuestionUpdated`/`Export.StudentsDownloaded` — bular hali
 *    `AuditActions.cs`da yo'q (AI sozlamalari va katalog UI'lari parallel/boshqa promptlarda
 *    yozilmoqda). Bu qiymatlar filtrga OLDINDAN qo'shilgan: agar hali yozuv yo'q bo'lsa filtr
 *    shunchaki bo'sh natija qaytaradi — xato emas.
 */
export interface AuditActionOption {
  value: string;
  labelKey: string;
}

export const AUDIT_ACTION_OPTIONS: readonly AuditActionOption[] = [
  { value: 'Auth.LoginSucceeded', labelKey: 'audit.actionLabels.AuthLoginSucceeded' },
  { value: 'Auth.LoginFailed', labelKey: 'audit.actionLabels.AuthLoginFailed' },
  { value: 'Auth.PasswordChanged', labelKey: 'audit.actionLabels.AuthPasswordChanged' },
  {
    value: 'Auth.TotpEnrollmentStarted',
    labelKey: 'audit.actionLabels.AuthTotpEnrollmentStarted',
  },
  { value: 'Auth.TotpEnabled', labelKey: 'audit.actionLabels.AuthTotpEnabled' },
  { value: 'Auth.TotpDisabled', labelKey: 'audit.actionLabels.AuthTotpDisabled' },
  { value: 'Security.RefreshReuse', labelKey: 'audit.actionLabels.SecurityRefreshReuse' },
  { value: 'School.Created', labelKey: 'audit.actionLabels.SchoolCreated' },
  { value: 'School.Updated', labelKey: 'audit.actionLabels.SchoolUpdated' },
  { value: 'School.Deleted', labelKey: 'audit.actionLabels.SchoolDeleted' },
  { value: 'School.LinkRegenerated', labelKey: 'audit.actionLabels.SchoolLinkRegenerated' },
  { value: 'School.ToggledActive', labelKey: 'audit.actionLabels.SchoolToggledActive' },
  { value: 'Student.Deleted', labelKey: 'audit.actionLabels.StudentDeleted' },
  { value: 'Assessment.Deleted', labelKey: 'audit.actionLabels.AssessmentDeleted' },
  {
    value: 'Assessment.ScoresRecalculated',
    labelKey: 'audit.actionLabels.AssessmentScoresRecalculated',
  },
  { value: 'Assessment.AnalysisRerun', labelKey: 'audit.actionLabels.AssessmentAnalysisRerun' },
  { value: 'Program.Created', labelKey: 'audit.actionLabels.ProgramCreated' },
  { value: 'Program.Updated', labelKey: 'audit.actionLabels.ProgramUpdated' },
  { value: 'Program.Published', labelKey: 'audit.actionLabels.ProgramPublished' },
  { value: 'Program.Archived', labelKey: 'audit.actionLabels.ProgramArchived' },
  { value: 'Program.Restored', labelKey: 'audit.actionLabels.ProgramRestored' },
  { value: 'Program.ToggledActive', labelKey: 'audit.actionLabels.ProgramToggledActive' },
  { value: 'Program.TestAdded', labelKey: 'audit.actionLabels.ProgramTestAdded' },
  { value: 'Program.TestRemoved', labelKey: 'audit.actionLabels.ProgramTestRemoved' },
  { value: 'Program.TestsReordered', labelKey: 'audit.actionLabels.ProgramTestsReordered' },
  { value: 'Program.Assigned', labelKey: 'audit.actionLabels.ProgramSchoolAssigned' },
  { value: 'Program.Unassigned', labelKey: 'audit.actionLabels.ProgramSchoolUnassigned' },
  { value: 'AiConfig.Updated', labelKey: 'audit.actionLabels.AiConfigUpdated' },
  { value: 'AiConfig.KeyChanged', labelKey: 'audit.actionLabels.AiConfigKeyChanged' },
  { value: 'Catalog.QuestionUpdated', labelKey: 'audit.actionLabels.CatalogQuestionUpdated' },
  // 2026-09-23 (`docs/07` §3.4.1): test biriktirmasi — ommaviy/maktablar/ommaviy makon/rejim.
  {
    value: 'Catalog.TestAssignmentUpdated',
    labelKey: 'audit.actionLabels.CatalogTestAssignmentUpdated',
  },
  { value: 'Export.StudentsDownloaded', labelKey: 'audit.actionLabels.ExportStudentsDownloaded' },
];

export interface AuditEntityTypeOption {
  value: string;
  labelKey: string;
}

export const AUDIT_ENTITY_TYPE_OPTIONS: readonly AuditEntityTypeOption[] = [
  { value: 'School', labelKey: 'audit.entityTypeLabels.School' },
  { value: 'Student', labelKey: 'audit.entityTypeLabels.Student' },
  { value: 'Assessment', labelKey: 'audit.entityTypeLabels.Assessment' },
  { value: 'AssessmentProgram', labelKey: 'audit.entityTypeLabels.AssessmentProgram' },
  { value: 'TestDefinition', labelKey: 'audit.entityTypeLabels.TestDefinition' },
  { value: 'AiProviderConfig', labelKey: 'audit.entityTypeLabels.AiProviderConfig' },
];

/** Harakat kodini (masalan `School.Created`) o'zbekcha nomga o'giradi; noma'lum bo'lsa xom kodni qaytaradi. */
export function findAuditActionLabelKey(action: string): string | undefined {
  return AUDIT_ACTION_OPTIONS.find((option) => option.value === action)?.labelKey;
}

/** Obyekt turini o'zbekcha nomga o'giradi; noma'lum bo'lsa xom qiymatni qaytaradi. */
export function findAuditEntityTypeLabelKey(entityType: string): string | undefined {
  return AUDIT_ENTITY_TYPE_OPTIONS.find((option) => option.value === entityType)?.labelKey;
}
